### PR TITLE
__mutable

### PR DESCRIPTION
**Pull request by tgehr**
Implementation of `__mutable` from the DConf hackathon.

### CODE REVIEW

**Comment by leandro-lucarella-sociomantic** [comment]
Hi, I'm not trying to be a PITA but am I the only one that think adding `__mutable` to the language if a profoundly bad idea?

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, a change in the language that is this big certainly needs a DIP, right?

**Comment by tgehr** [comment]
It does and I currently have only a rough draft, but Walter told me to create this pull request anyway.

**Comment by tgehr** [comment]
For some reason my first comment is not here anymore, so again: this is not C++ `mutable`, `__mutable` is unsafe, because it does not disable rewrites based on `immuable`/`const`/`pure`. The reason this is being suggested at all is that Andrei wants the language to be able to store allocators within `immutable` containers. I don't think there is a better solution.

**Comment by tgehr** [comment]
Also @leandro-lucarella-sociomantic : Maybe put forward some arguments in favor of your position. Do you think it is unreasonable to store references to allocators in `immutable` collections, or do you think there is a better solution?

File: src/dmd/mtype.d [hunk 1]
```diff
@@ -1961,6 +1961,38 @@ extern (C++) abstract class Type : RootObject
         return t;
     }
 
+    /************************************
+     * Add MODxxxx bits to type of __mutable member variable
```
Comment (WalterBright) [inline src/dmd/mtype.d]:
```
Returns:
    type with added bits
```

**Review by WalterBright — APPROVED** [review]

**Comment by leandro-lucarella-sociomantic** [comment]
I agree with @WalterBright that is better to leave the discussion about language change to the DIP and their review process, and keep only discussion about the implementation here. So I will follow that.

I think when submitting PRs for a new language feature that doesn't even have a DIP, it would be nice to make it clear to avoid hard attacks on some people :smiley: 

Like having a `[Proposal] Draft implementation of __mutable` in the title and something like `This PR is a reference implementation for an upcoming DIP, please don't merge yet` in the issue description.

File: test/compilable/__mutable.d [hunk 2]
```diff
@@ -0,0 +1,92 @@
+// REQUIRED_ARGS: -o-
```
Comment (wilzbach) [inline test/compilable/__mutable.d]: Why disable the object generation?
I think it wouldn't hurt to have at least one test that does test that `__mutable` interacts fine without the backend failing with it.

File: src/dmd/traits.d [hunk 3]
```diff
@@ -699,6 +700,14 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
 
         return isDeclX(d => (d.storage_class & STC.lazy_) != 0);
     }
+    if (e.ident == Id.isUnderUnderMutable)
+    {
+        if (dim != 1)
+            return dimError(1);
```
Comment (wilzbach) [inline src/dmd/traits.d]: Not covered.

File: src/dmd/dsymbolsem.d [hunk 4]
```diff
@@ -1036,6 +1036,29 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
             }
         }
 
+        if (dsym.storage_class & STC.mutable)
+        {
+            if (!dsym.isField())
+            {
+                dsym.error("only fields can be `__mutable`");
```
Comment (wilzbach) [inline src/dmd/dsymbolsem.d]: This results in hard to understand error message:

```
Error: variable `__mutable2.foo.p` only fields can be `__mutable`
```

maybe "cannot be __mutable. Only fields can."?

File: src/dmd/dsymbolsem.d [hunk 5]
```diff
@@ -1036,6 +1036,29 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
...
+                {
+                    dsym.error("cannot be both `__mutable` and `immutable`");
+                }
+                if (dsym.storage_class & STC.const_)
+                {
+                    dsym.error("cannot be both `__mutable` and `const`");
+                }
+                if (dsym.protection.kind != Prot.Kind.private_)
+                {
+                    dsym.error("`__mutable` fields must be `private`");
```
Comment (wilzbach) [inline src/dmd/dsymbolsem.d]: Same thing here about the error message:

```
Error: variable `__mutable1.S.pm` `__mutable` fields must be `private`
```

maybe "can only be `private`"? Or triggering a full error with: "__mutable field X can only be private"?

File: src/dmd/mtype.d (hunk 1 above)
Comment (ntrel) [reply src/dmd/mtype.d]: Redundancy in documentation is sometimes useful, e.g. if the `Returns` information was also mentioned in a *long* description, so the return can be quickly looked up. Here, reading the *short* description is clearly enough IMO.

File: src/dmd/dsymbolsem.d (hunk 4 above)
Comment (bausshf) [reply src/dmd/dsymbolsem.d]: Or maybe "cannot be declared `__mutable`. `__mutable` can only be applied to fields."

**Comment by RazvanN7** [comment]
Closing, as this endeavor is not going to be pursued.

**Pull request closed without merging**